src/main/java/network/brightspots/rcv/ResultsWriter.java

```diff
@@ -87,49 +89,131 @@ class ResultsWriter {
+    }
+
+    public String getBasename() {
+      return basename;
+    }
+
+    public String getExtension() {
+      return extension;
+    }
   }
 
-  // visible for testing
-  @SuppressWarnings("WeakerAccess")
-  static String getOutputFilePath(
-      String outputDirectory,
-      String outputType,
-      String timestampString,
-      String sequentialTabulationId) {
-    outputType = sanitizeStringForOutput(outputType);
-    String fileName =
-        String.format(
-            "%s_%s%s",
-            timestampString, outputType, sequentialSuffixForOutputPath(sequentialTabulationId));
-    return Paths.get(outputDirectory, fileName).toAbsolutePath().toString();
+  public static class ResultFile {
+    // Since sanitizing Slice IDs can cause filename collisions, ensure each non-sanitized
+    // Slice ID is given a unique sanitized name.
+    private static final Dictionary<String, String> sliceIdToUniqueSanitizedId = new Hashtable<>();
+    // This is a set of the values in sanitizerCollisionResolution to ensure there are no collisions
+    // in the resulting sanitized names.
+    private static final HashSet<String> uniqueSanitizedIds = new HashSet<>();
+    private final ResultType resultType;
+    private final boolean isSlice;
+    private final TabulateBySlice slice;
+    private final String sliceId;
+
+    ResultFile(ResultType resultType) {
+      this.resultType = resultType;
+      this.isSlice = false;
+      slice = null;
+      sliceId = null;
+    }
+
+    ResultFile(ResultType resultType, TabulateBySlice slice, String sliceId) {
+      this.resultType = resultType;
+      this.isSlice = true;
+      this.slice = slice;
+      this.sliceId = sliceId;
+    }
+
+    // getPath helper without the modifier argument
+    public Path getPath(String directory, String prefix, Integer sequentialId) {
+      return getPath(directory, prefix, null, sequentialId);
+    }
+
+    // The filename is an underscore-separated string containing all components needed
+    // to create a unique output file. The modifier is likely only used in tests, where
+    // it is set to "expected" to distinguish expected output files from actual output files.
+    public Path getPath(String directory, String prefix, String modifier, Integer sequentialId) {
+      List<String> parts = new ArrayList<>();
+      parts.add(prefix);
+      if (sequentialId != null) {
+        parts.add(sequentialId.toString());
+      }
+      if (modifier != null) {
+        parts.add(modifier);
+      }
+      if (isSlice) {
+        assert slice != null;
+        parts.add(sanitizeSliceWithoutCollisions(sliceId));
+        parts.add(slice.toLowerString());
+      }
+      parts.add(resultType.getBasename());
+
+      String filenameWithoutExt = String.join("_", parts);
+      return Path.of(directory, "%s.%s".formatted(filenameWithoutExt, resultType.getExtension()));
+    }
+
+    private String sanitizeSliceWithoutCollisions(String sliceId) {
```

Comment:
I'm afraid to give any of our code the appearance of being thread-safe (other than the UI). For compliance, we generally can't do anything multithreaded because we need deterministic outputs.

For example, if we use a random seed as a tiebreaker, the RNG needs to be called in precisely the same order every tabulation run.

Reply:
I think it's possible to have deterministic outputs in a multi-threaded environment, but it also sounds like I've got the wrong reviewer hat on for this one! If it's guaranteed that this code isn't accessed from multiple threads, then I have no concern.